core/unit_test/TestArrayOps.hpp

```diff
@@ -0,0 +1,491 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#if defined(__clang__)
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wchar-subscripts"
+#endif
+
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wchar-subscripts"
+#endif
```

Comment:
Why do we bother with accesses with all integral types known to men?

Reply:
I'd like to eventually change it to `is_convertible_v<T, size_t>` (which is closer to how `std::array` works).  Because it has been around a long time, I don't feel comfortable making that change w/o a full set of tests to make sure nothing is missing or generates other warnings.

`std::array` just uses `size_t`.  I'm assuming the `is_integral` was used in `Array` to eliminate compiler warnings that the implicit conversion sometimes generates.

Reply:
Could `array_element_access` test be wrapped into a type-parametrized test?
https://github.com/google/googletest/blob/main/docs/advanced.md#type-parameterized-tests

Reply:
Ok, I couldn't resist and tried to rewrite it on my own. Consider https://github.com/cz4rs/kokkos/commit/be291da82ce6146bb1a832d525ba3b9b1f8f2a4c, which would produce following test output:
```
Note: Google Test filter = *ArrayOpsTest*
[==========] Running 24 tests from 12 test suites.
[----------] Global test environment set-up.
[----------] 2 tests from serial/ArrayOpsTest/0, where TypeParam = signed char
[ RUN      ] serial/ArrayOpsTest/0.array_element_access
[       OK ] serial/ArrayOpsTest/0.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/0.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/0.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/0 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/1, where TypeParam = unsigned char
[ RUN      ] serial/ArrayOpsTest/1.array_element_access
[       OK ] serial/ArrayOpsTest/1.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/1.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/1.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/1 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/2, where TypeParam = short
[ RUN      ] serial/ArrayOpsTest/2.array_element_access
[       OK ] serial/ArrayOpsTest/2.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/2.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/2.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/2 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/3, where TypeParam = unsigned short
[ RUN      ] serial/ArrayOpsTest/3.array_element_access
[       OK ] serial/ArrayOpsTest/3.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/3.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/3.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/3 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/4, where TypeParam = int
[ RUN      ] serial/ArrayOpsTest/4.array_element_access
[       OK ] serial/ArrayOpsTest/4.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/4.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/4.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/4 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/5, where TypeParam = unsigned int
[ RUN      ] serial/ArrayOpsTest/5.array_element_access
[       OK ] serial/ArrayOpsTest/5.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/5.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/5.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/5 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/6, where TypeParam = long
[ RUN      ] serial/ArrayOpsTest/6.array_element_access
[       OK ] serial/ArrayOpsTest/6.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/6.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/6.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/6 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/7, where TypeParam = unsigned long
[ RUN      ] serial/ArrayOpsTest/7.array_element_access
[       OK ] serial/ArrayOpsTest/7.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/7.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/7.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/7 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/8, where TypeParam = long long
[ RUN      ] serial/ArrayOpsTest/8.array_element_access
[       OK ] serial/ArrayOpsTest/8.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/8.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/8.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/8 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/9, where TypeParam = unsigned long long
[ RUN      ] serial/ArrayOpsTest/9.array_element_access
[       OK ] serial/ArrayOpsTest/9.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/9.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/9.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/9 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/10, where TypeParam = (anonymous namespace)::Enum
[ RUN      ] serial/ArrayOpsTest/10.array_element_access
[       OK ] serial/ArrayOpsTest/10.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/10.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/10.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/10 (0 ms total)

[----------] 2 tests from serial/ArrayOpsTest/11, where TypeParam = (anonymous namespace)::EnumShort
[ RUN      ] serial/ArrayOpsTest/11.array_element_access
[       OK ] serial/ArrayOpsTest/11.array_element_access (0 ms)
[ RUN      ] serial/ArrayOpsTest/11.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/11.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/11 (0 ms total)

[----------] Global test environment tear-down
[==========] 24 tests from 12 test suites ran. (0 ms total)
[  PASSED  ] 24 tests.
```

Reply:
Interesting...

The error messages look fine as well (as they show what `TypeParam` is on the FAILED line:

```
[----------] 2 tests from serial/ArrayOpsTest/0, where TypeParam = signed char
[ RUN      ] serial/ArrayOpsTest/0.array_element_access
/Users/nliber/git/github.com/kokkos/kokkos-array/core/unit_test/TestArrayOps.hpp:48: Failure
Expected: (a[cast_index]) != (a[index]), actual: 5 vs 5
[  FAILED  ] serial/ArrayOpsTest/0.array_element_access, where TypeParam = signed char (0 ms)
[ RUN      ] serial/ArrayOpsTest/0.array_contiguous_element_access
[       OK ] serial/ArrayOpsTest/0.array_contiguous_element_access (0 ms)
[----------] 2 tests from serial/ArrayOpsTest/0 (0 ms total)
```

Reply:
Unfortunately, some of the GoogleTest macros involved with the type parameterized tests don't work across our compiler suite.  Reverting...

Reply:
Oh, I've never tried to do full CI run :| In that case I think this test is good as it is.